### DISCUSSION TITLE
Publishing on Steam Solutions

### DISCUSSION THREAD

**Discussion by RossD20Studios** [airsdk/Adobe-Runtime-Support#1089] [Guides & Tutorials]
Hello fellow AIR Runtime users - 

I'm posting solutions here to help those publishing their AIR games to Steam.

1. Start by following the Steamworks documentation to configure your build depots for each target platform and associating those build depots with your application packages.
2. If your app is < 2GB, you can upload the build as a zip file containing the application output - otherwise, you need to use the SteamPipe to upload.
3. If you're uploading a Mac build, it's necessary to strip the game of the wrapper that gets added during the build upload process. To do this, you need to use the ContentPre.app in Steamworks SDK. You will need to run this from the Mac terminal with the following parameters:

`ContentPrep.app/Contents/MacOS/contentprep.py --nowrap --console -v -s MyGame.app -d output -a appId -e MyGame.app/Contents/MacOS/MyGame`

Replace "MyGame" with your application name.
Replace appId with integer of your Steam application id.
**Note:** The ContentPrep.app will not run as a GUI, so the --console command is needed or you'll get errors about needing to install an outdated version of wxPython (Special thanks to Alan Hazelden who identified this fix)

4. Configure your launch options for each platform in Installation -> General Installation. For windows build, use the game's EXE. For mac build, use MyGame.app.
5. Publish and test launch via Steam client before submitting to Steam for review. If your game doesn't launch through Steam client, they'll reject it in the review.

Hope this helps those of you publishing games to Steam! Let me know if you run into any trouble or need additional details to get your AIR game running on Steam.

~Ross

**Comment by marchbold** [comment]
Awesome information!

**Reply by RossD20Studios** [reply]
Thanks! I haven't tried the SteamPipe yet, but I gather from reading the Steam articles and talking with other Steam devs that this also fixes the Mac application so it runs correctly. You need to setup the build scripts to use this, so it's a little more work up front than just uploading the ZIP, but I imagine it might be a better long-term solution. I'll eventually give this a try and report back if it works.

Are you also developing games with AIR for Steam? I'd love to connect with other devs doing similar work and see what kind of projects folks are creating these days with AIR.

**Reply by marchbold** [reply]
I don't have any games on Steam, however I'm currently integrating our [GameServices ANE](https://airnativeextensions.com/extension/com.distriqt.GameServices) to provide access to some of the Steam functionality (achievements / leaderboards etc) so I'm playing around with that side of it. 

So yeah I'm also very interested to hear dev's experience with the platform.

**Reply by RossD20Studios** [reply]
That's fantastic news! I'm very interested to get Steam achievements in our game. I had been looking at [FRESteamWorks](https://github.com/Ventero/FRESteamWorks) but the last code update on that was over four years ago. I see you already support Google Play and Apple, so I take it that your ANE code streamlines the API's so developers can program a single achievement system that works across all platforms? Also, I currently use your In App Billing extension. Do you have plans to add Steam support to this extension as well?

Please let me know if you'd like any help testing your Steam integrations. My game project is called [Summoners Fate](https://store.steampowered.com/app/1681840/Summoners_Fate/). I'd be happy to try integrating your ANE into our game, get you build access, and work through any issues together.

**Reply by waveofthought-code** [reply]
>I had been looking at [FRESteamWorks](https://github.com/Ventero/FRESteamWorks) but the last code update on that was over four years ago.

fwiw this ANE works fine still but you have to rebuild it in xcode for 64bit if you want to release for macos.   Achievements are supported

**Reply by marchbold** [reply]
@RossD20Studios That's the idea, the steam integration will drop in as a service on macOS alongside game center and as the only service on windows. Same API across the services and platforms.

Definitely will need help testing soon, so would really appreciate any help! I'll reach out when we have something to test, or shoot us an email at airnativeextensions@distriqt.com and I'll arrange access to the beta.

@atmofunk Yeah aware that still works, and definitely nothing wrong with using it, we just want to provide a complete solution across several services to make life easier for devs.

**Reply by jimmymjing** [reply]
@marchbold Hello, I would like to test Steam functionality for GameServices ANE on our game. Is it possible to access the beta version?

**Reply by marchbold** [reply]
@jimmymjing Unfortunately the beta has been delayed. We've had a lot of interruptions with our work lately (holidays + covid + other illnesses) and development projects like this have been put on hold. We are hoping to get back into them soon.

**Reply by jimmymjing** [reply]
Oh sure! Please count me in once it's back on track~

**Comment by KupoGames** [comment]
I'd like to add some Steam tips of my own, just in case anyone still sees this thread:

1. The FRESteamWorks ANE is great for adding Steam achievements and leaderboards, and is really easy to use.
2. The Steam overlay won't work correctly (it'll freeze) if your app's graphics aren't being updated, so make sure there's some animation going on at all times.
3. Even ActionScript 2 games can be published on Steam! Simply use AIR to open the swf using HTMLloader, and communicate with it using LocalConnection. (I'm sure there's multiple ways of doing this, that's just my approach)

So far I've published Epic Battle Fantasy 3 (an AS2 game), EBF4 and EBF5 (AS3 games), and Bullet Heaven 2 (an AS3 game made with Starling). AIR/Flash is totally still a viable game engine for Desktop, especially so if you use a framework like Starling to get better performance.

**Comment by FliplineStudios** [comment]
Hi everyone, we're a longtime iOS/Android game developer and starting to look into releasing desktop games on Steam.  As we're looking through the processes of packaging for Windows and uploading to Steam, I had a question about signing our builds.  

In the past we've used a self-signed certificate for Android builds, and of course Apple's certificates for iOS builds.  For releasing a Windows game via Steam, do we need to get an official certificate (like from digicert or elsewhere), or do we continue to use a self-signed cert like we do for Android?  Curious about whether Windows or antivirus software will complain about a self-cert when distributed on Steam, and if Valve has any requirements on this (I haven't noticed anything in the docs so far).

If anyone has any info about this, and/or suggestions for the type of cert needed (EV, etc.) and any recommendations, we'd appreciate it!

**Reply by KupoGames** [reply]
I also use a self-signed certificate. Basically, as long as your app is not a virus, that's legit enough for Steam!
I have had some antivirus programs occasionally complain about my app though - but that's because I use another application, Qres, to change screen resolutions in an old version of AIR. I think this isn't needed any more in newer versions of AIR which have more Desktop resolution options. Anyway, I only got these complaints when my app first launched, and not too long later the antiviruses stopped complaining about it.

**Comment by RossD20Studios** [comment]
I use a self-signed cert for my Steam game like I do for Android. Valve
does not have any requirements that I'm aware of. On rare occasions, I have
had users report that their virus protection software flags the game.
